### PR TITLE
Ann/issue447

### PR DESCRIPTION
**Pull request by WeiWW**

### CODE REVIEW

**Comment by mTwTm** [comment]
@walkingice can you help review this? I'm not familiar with our `RelocateService`.

File: app/src/main/java/org/mozilla/focus/download/DownloadInfoManager.java [hunk 1]
```diff
@@ -92,7 +93,17 @@ protected void onQueryComplete(int token, Object cookie, Cursor cursor) {
                         List<DownloadInfo> downloadInfoList = new ArrayList<>();
                         if (cursor != null) {
                             while (cursor.moveToNext()) {
-                                downloadInfoList.add(cursorToDownloadInfo(cursor));
+                                DownloadInfo downloadInfo = cursorToDownloadInfo(cursor);
+                                File file = new File(URI.create(downloadInfo.getFileUri()).getPath());
+
+                                if (file.exists()){
+                                    downloadInfoList.add(downloadInfo);
+                                }else {
```
Comment (walkingice) [inline app/src/main/java/org/mozilla/focus/download/DownloadInfoManager.java]: white space before "else"

File: app/src/main/java/org/mozilla/focus/download/DownloadInfoManager.java [hunk 2]
```diff
@@ -92,7 +93,17 @@ protected void onQueryComplete(int token, Object cookie, Cursor cursor) {
                         List<DownloadInfo> downloadInfoList = new ArrayList<>();
                         if (cursor != null) {
                             while (cursor.moveToNext()) {
-                                downloadInfoList.add(cursorToDownloadInfo(cursor));
+                                DownloadInfo downloadInfo = cursorToDownloadInfo(cursor);
```
Comment (walkingice) [inline app/src/main/java/org/mozilla/focus/download/DownloadInfoManager.java]: final

File: app/src/main/java/org/mozilla/focus/download/DownloadInfoManager.java [hunk 3]
```diff
@@ -174,6 +174,11 @@ public void query(int offset, int limit, AsyncQueryListener listener) {
         mQueryHandler.startQuery(TOKEN, listener, Uri.parse(uri), null, null, null, null);
     }
 
+    public void queryCertainId(Long downloadId,AsyncQueryListener listener){
+        String uri = Download.CONTENT_URI.toString();
+        mQueryHandler.startQuery(TOKEN,listener,Uri.parse(uri),null,Download.DOWNLOAD_ID+"==?",new String[] {Long.toString(downloadId)},null);
```
Comment (walkingice) [inline app/src/main/java/org/mozilla/focus/download/DownloadInfoManager.java]: just use `Download.CONTENT_URI` ?

**Comment by mTwTm** [comment]
Fixes #447, #459

File: app/src/main/java/org/mozilla/focus/download/DownloadInfoManager.java (hunk 3 above)
Comment (WeiWW) [reply app/src/main/java/org/mozilla/focus/download/DownloadInfoManager.java]: YES

File: app/src/main/java/org/mozilla/focus/fragment/DownloadsFragment.java [hunk 4]
```diff
@@ -53,7 +52,9 @@ public void onReceive(Context context, Intent intent) {
                         public void onQueryComplete(List downloadInfoList) {
                             for (int i=0;i<downloadInfoList.size();i++){
                                 DownloadInfo downloadInfo = (DownloadInfo) downloadInfoList.get(i);
-                                mDownloadListAdapter.updateItem(downloadInfo);
+                                if (mDownloadListAdapter != null){
```
Comment (walkingice) [inline app/src/main/java/org/mozilla/focus/fragment/DownloadsFragment.java]: The BroadcastReceiver is created after creation of DownloadListAdapter, how come the adapter becomes null?

File: app/src/main/java/org/mozilla/focus/fragment/DownloadsFragment.java [hunk 5]
```diff
@@ -30,18 +34,31 @@ public static DownloadsFragment newInstance() {
     public View onCreateView(LayoutInflater inflater, ViewGroup container,
                              Bundle savedInstanceState) {
         recyclerView = (RecyclerView) inflater.inflate(R.layout.fragment_downloads, container, false);
+        mDownloadListAdapter = new DownloadListAdapter(recyclerView.getContext());
```
Comment (walkingice) [inline app/src/main/java/org/mozilla/focus/fragment/DownloadsFragment.java]: In `onCreate`, the BroadcastReceiver is created after creation of DownloadListAdapter, how come the adapter becomes null?

**Pull request merged**